### PR TITLE
Merging pmaytak/perf

### PR DESCRIPTION
**Pull request by pmaytak**
Merges only test related changes. Merge perf client and benchmark project. Includes Henrik's test runner changes.

Note: Will fix integration tests after merge.

#88

### CODE REVIEW

**Review by jennyf19 — CHANGES_REQUESTED** [review]
:clock1:

**Comment by jennyf19** [comment]
will wait to review when integration tests are fixed.

**Comment by pmaytak** [comment]
@jennyf19 Integration test is fixed. I moved perf related code from IntegrationTestService to new project PerformanceTestService.

File: tests/IntegrationTests/IntegrationTestService/IntegrationTestService.csproj [hunk 1]
```diff
@@ -1,12 +1,11 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk.Web">
 
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp3.1; net5.0</TargetFrameworks>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
     <IsPackable>false</IsPackable>
```
Comment (jennyf19) [inline tests/IntegrationTests/IntegrationTestService/IntegrationTestService.csproj:5]: nit: can it still run on net 5? i don't remember.

File: tests/IntegrationTests/WebAppUiTests/WebAppUiTests.csproj [hunk 2]
```diff
@@ -1,7 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp3.1; net5.0</TargetFrameworks>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
Comment (jennyf19) [inline tests/IntegrationTests/WebAppUiTests/WebAppUiTests.csproj:4]: can net5 still be used?

File: tests/Microsoft.Identity.Web.Test.Common/TestConstants.cs [hunk 3]
```diff
@@ -124,7 +124,12 @@ public static class TestConstants
         public const string KeyVaultReference = "AzureADIdentityDivisionTestAgentCert";
 
         // Integration tests
-        public const string SecurePageGetTokenAsync = "/SecurePage/GetTokenAsync";
+        public const string EmptyGetEmpty = "/Empty";
+        public const string TokenAcquisitionGetEmpty = "/TokenAcquisition";
+        public const string GraphClientGetEmpty = "/GraphClient";
+        public const string SecurePageGetEmpty = "/SecurePage/GetEmpty";
```
Comment (jennyf19) [inline tests/Microsoft.Identity.Web.Test.Common/TestConstants.cs:130]: nit: should this be "/GraphClient/GetEmpty" or change the variable name to something else?

File: tests/Microsoft.Identity.Web.Test.Integration/Microsoft.Identity.Web.Test.Integration.csproj [hunk 4]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-     <TargetFrameworks>netcoreapp3.1; net5.0</TargetFrameworks>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
Comment (jennyf19) [inline tests/Microsoft.Identity.Web.Test.Integration/Microsoft.Identity.Web.Test.Integration.csproj:4]: same as other net5 question

File: tests/PerformanceTests/Microsoft.Identity.Web.Perf.Benchmark/TokenAcquisitionTests.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+        [Benchmark]
+        public void GetAccessTokenForAppAsync()
+        {
+            HttpRequestMessage httpRequestMessage = new HttpRequestMessage(HttpMethod.Get, TestConstants.SecurePageGetTokenForAppAsync);
+            HttpResponseMessage response = _client.SendAsync(httpRequestMessage).GetAwaiter().GetResult();
+            if (!response.IsSuccessStatusCode)
+            {
+                throw new Exception("Failed.");
+            }
```
Comment (jennyf19) [inline tests/PerformanceTests/Microsoft.Identity.Web.Perf.Benchmark/TokenAcquisitionTests.cs:75]: do you want to return the error code? maybe include which test this is for?

File: tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/appsettings.json [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{
+    "IntegrationTestServicesBaseUri": "https://msidwebintegrationtestservice.azurewebsites.net",
+    "TenantDomain": "msidwebperformancetesting.onmicrosoft.com",
+    "DurationInMinutes": "1440",
+    "UserPassword": "secret_here",
+    "TestUri": "/SecurePage/GetTokenForUserAsync",
```
Comment (jennyf19) [inline tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/appsettings.json:6]: is there a way for this to run auto-magically in vsts w/the other tests or run continuously? is it possible to get the lab to implement something similar so we can use them to get the password and have it run as often as needed?

File: tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/appsettings.json [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+    "ApiScopes": "api://b05c7590-7c68-4aba-9454-38b1f787a989/.default",
+    "ClientId": "91f7fafd-e19a-4637-b3b9-1550fb132469",
+    "UsersStartIndex": "1", //start w/at least one
+    "UsersEndIndex": "10000"
+}
```
Comment (jennyf19) [inline tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/appsettings.json:11]: nit: maybe add a note that we only have 10k users. :)

File: tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/ScalableTokenCacheHelper.cs [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+using System.Threading;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Identity.Client;
+using Newtonsoft.Json;
+
+namespace Microsoft.Identity.Web.Perf.Client
+{
+    /// <summary>
+    /// Token cache writing on disk one cache per account
+    /// WARNING: this version is not encrypted
```
Comment (jennyf19) [inline tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/ScalableTokenCacheHelper.cs:20]: >WARNING: [](start = 8, length = 8)

lol.

**Comment by jennyf19** [comment]
> 

I assume this is the same as what @jmprieur committed...i just skipped reviewing it as I had previously when i pulled his changes.

---
Refers to: tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/ScalableTokenCacheHelper.cs:201 in aa4a04f. [](commit_id = aa4a04f2ae597b2c4aa68cdb4c1be85f905f013c, deletion_comment = False)

File: tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/TestRunner.cs [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+                                else
+                                {
+                                    if (i % 10 == 0)
+                                    {
+                                        ScalableTokenCacheHelper.PersistCache();
+                                    }
+                                }
+
+                                // Console.Write(".");
+
```
Comment (jennyf19) [inline tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/TestRunner.cs:119]: nit: remove?

**Comment by jennyf19** [comment]
> using Microsoft.AspNetCore.Authorization;

license info.

---
Refers to: tests/PerformanceTests/PerformanceTestService/Controllers/EmptyController.cs:1 in aa4a04f. [](commit_id = aa4a04f2ae597b2c4aa68cdb4c1be85f905f013c, deletion_comment = False)

**Comment by jennyf19** [comment]
> using Microsoft.AspNetCore.Authorization;

license info.

---
Refers to: tests/PerformanceTests/PerformanceTestService/Controllers/GraphClientController.cs:1 in aa4a04f. [](commit_id = aa4a04f2ae597b2c4aa68cdb4c1be85f905f013c, deletion_comment = False)

**Comment by jennyf19** [comment]
> using Microsoft.AspNetCore.Authorization;

license info.

---
Refers to: tests/PerformanceTests/PerformanceTestService/Controllers/TokenAcquisitionController.cs:1 in aa4a04f. [](commit_id = aa4a04f2ae597b2c4aa68cdb4c1be85f905f013c, deletion_comment = False)

**Comment by jennyf19** [comment]
there were some changes made to the msalInMemoryTokenCacheProvider & one other class in my previous branch, but i'll go back and add those in after you merge.

---
In reply to: [702424303](https://github.com/AzureAD/microsoft-identity-web/pull/636#issuecomment-702424303) [](ancestors = 702424303)

File: tests/IntegrationTests/IntegrationTestService/IntegrationTestService.csproj (hunk 1 above)
Comment (jennyf19) [reply tests/IntegrationTests/IntegrationTestService/IntegrationTestService.csproj:5]: nevermind, i removed it, i just don't remember why right now. 

---
In reply to: [498561988](https://github.com/AzureAD/microsoft-identity-web/pull/636#discussion_r498561988) [](ancestors = 498561988)

**Review by jennyf19 — APPROVED** [review]
:shipit:

File: tests/IntegrationTests/IntegrationTestService/IntegrationTestService.csproj (hunk 1 above)
Comment (pmaytak) [reply tests/IntegrationTests/IntegrationTestService/IntegrationTestService.csproj:5]: :)

File: tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/appsettings.json (hunk 6 above)
Comment (pmaytak) [reply tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/appsettings.json:6]: Yea, I'm sure there's a way. Maybe Azure Functions or WebJob.

File: tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/TokenCacheHelper.cs [hunk 10]
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (jmprieur) [inline tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/TokenCacheHelper.cs:1]: Do we needs this class?

File: tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/appsettings.json (hunk 7 above)
Comment (jmprieur) [reply tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/appsettings.json:11]: It"s a console app.
If it gets its creds from KeyVault, you can run it with a release definition (used for tests), triggered by the CI Build

File: tests/PerformanceTests/PerformanceTestService/Controllers/WeatherForecastController.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Identity.Web;
+using Microsoft.Identity.Web.Test.Common;
+
+namespace PerformanceTestService.Controllers
+{
+    [ApiController]
+    [Authorize]
+    [Route("SecurePage")]
+    public class WeatherForecastController : ControllerBase
```
Comment (jmprieur) [inline tests/PerformanceTests/PerformanceTestService/Controllers/WeatherForecastController.cs:15]: We probably want to rename the file and the class tio SecurePageController ?

File: tests/PerformanceTests/PerformanceTestService/EventSource/BenchmarkMsalDistributedTokenCacheAdapter.cs [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Options;
+using Microsoft.Identity.Web.TokenCacheProviders.Distributed;
+
+namespace PerformanceTestService.EventSource
+{
+    /// <summary>
+    /// Adds benchmarking counters on top of <see cref="MsalDistributedTokenCacheAdapter"/>.
+    /// </summary>
+    public class BenchmarkMsalDistributedTokenCacheAdapter : MsalDistributedTokenCacheAdapter
```
Comment (jmprieur) [inline tests/PerformanceTests/PerformanceTestService/EventSource/BenchmarkMsalDistributedTokenCacheAdapter.cs:14]: Is it used?

**Review by jmprieur — COMMENTED** [review]
@pmaytak : I propose a few changes

File: tests/PerformanceTests/PerformanceTestService/Controllers/WeatherForecastController.cs (hunk 11 above)
Comment (pmaytak) [reply tests/PerformanceTests/PerformanceTestService/Controllers/WeatherForecastController.cs:15]: Yea, I'll refactor the naming.

File: tests/PerformanceTests/PerformanceTestService/EventSource/BenchmarkMsalDistributedTokenCacheAdapter.cs (hunk 12 above)
Comment (pmaytak) [reply tests/PerformanceTests/PerformanceTestService/EventSource/BenchmarkMsalDistributedTokenCacheAdapter.cs:14]: Yes, it's used in place of normal distributed cache class.

File: tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/TokenCacheHelper.cs (hunk 10 above)
Comment (pmaytak) [reply tests/PerformanceTests/Microsoft.Identity.Web.Perf.Client/TokenCacheHelper.cs:1]: I'll verify if we need it.

**Pull request merged**